File: Algorithm/C++/1000+/1022. Sum of Root To Leaf Binary Numbers.cc
/*
 * Created on Tue Apr 09 2019 10:26:14
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
 public:
  int sumRootToLeaf(TreeNode* root) {
    return dfs(root, 0);
  }
 private:
  int dfs(TreeNode* root, int curr) {
    if (root->left == nullptr && root->right == nullptr) {
      return curr * 2 + root->val;
    }
    int tmp = 0;
    if (root->left != nullptr) tmp += dfs(root->left, curr * 2 + root->val);
    if (root->right != nullptr) tmp += dfs(root->right, curr * 2 + root->val);
    return tmp;
  }
};

File: Algorithm/Go/1000+/1022. Sum of Root To Leaf Binary Numbers.go
/*
 * Created on Tue Apr 09 2019 10:31:19
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func sumRootToLeaf(root *TreeNode) int {
	return dfs(root, 0)
}

func dfs(root *TreeNode, curr int) int {
	if root.Left == nil && root.Right == nil {
		return curr*2 + root.Val
	}
	tmp := 0
	if root.Left != nil {
		tmp += dfs(root.Left, curr * 2 + root.Val)
	}
	if root.Right != nil {
		tmp += dfs(root.Right, curr * 2 + root.Val)
	}
	return tmp
}

